Commonalities/documentation/Working/API-design-guidelines.md

```diff
@@ -0,0 +1,989 @@
+```http
+https://api.mycompany.com/v1/orders?order_by=rating,reviews,name&order=desc
+```
+
+### 8.3 Filtering
+
+
+Filtering consists of restricting the number of resources queried by specifying some attributes and their expected values. It is possible to filter a collection on multiple attributes at the same time and allow multiple values for a filtered attribute.
+
+Next, it is specified how it should be used according to the filtering based on the type of data being searched for: a text, a number or a date and the type of operation.
+ 
+| **Operation** |	Text |	Numbers | 	Dates |
+| ----- |	-----  |	 -----  |  -----  |
+| Equal | `GET .../?name=Juan` | `GET .../?amount=807.24`	| `GET .../?executionDate=2018-30-05` | 
+| Greater or equal	| N/A | `GET .../?fromAmount=807.24` | `GET.../?fromExecutionDate=2018-30-05` |
+| smaller or equal | N/A | `GET .../?toAmount=807.24` | `GET.../?toExecutionDate=2018-30-05` | 
+|Contains | `GET .../?name=~Juan` |N/A | N/A | 
+
+
+**Additional rules**:
+- The operator "`&`" is evaluated as an AND between different attributes.
+- A Query Param (attribute) can contain 1 or n values separated by "`,`".
+- For operations on numeric, date or enumerated fields, the use of the prefixes “from” and “to” will be allowed, which will act as comparators for “greater than or equal to” and “less than or equal to”.
+
+**Examples**:
+- <u>Equals</u>: to search for users with first name "david" and last name "munoz":
+  - `GET /users?name=david&surname=munoz`
+  - `GET /users?name=David,Noelia`
+    - Search for several values separating them by "`,`".
+- <u>Inclusion</u>: if we already have a filter that searches for "equal" and we want to provide it with the possibility of searching for "inclusion", we must include the character "~"
+  - `GET /users?name=dav`
+    - Search for the exact name "dav"
+  - `GET /users?name=~dav`
+    - Look for names that include "dav"
+- <u>Greater than / less than</u>: new attribute will be created and it will be preceded with a "from" and a "to".
+  - `GET /users?fromCreatrionDate=2021-01-01T00:00:00`
+    - Find users with creation Date greater than 2021
+  - `GET /users?toCreationDate=2021-11-31T23:59:59`
+    - Find users with creationDate less than 2022
+  - `GET /users?fromCreationDate=2020-01-01T00:00:00&toCreationDate=2021-11-31T23:59:59`
+    - Search for users created between 2020 and 2021
+
+
+## 9. Architecture Headers
+
+With the aim of standardizing the request observability and traceability process, common headers that provide a follow-up of the E2E processes should be included. The table below captures these headers.
+
+| Name | Description |  Type | Pattern	| Longitude | Required |	Example | 
+|---|---|---|---|---|---|---|  
+| `X-Version` |	Service version indentificator |	String| N/A	| | No | |	
+| `X-Correlator`|	Service correlator to make observability|		String |	UUID (8-4-4-4-12)	| Max 36	| Yes |	b4333c46-49c0-4f62-80d7-f0ef930f1c46 |
```

Comment:
I disagree with X-Correlator being mandatory, as API consumers often don't care and just set this to a fixed value. If they don't care, far better that they don't set anything, and let the network do that to ensure a unique value for tracing purposes. Rejecting duplicate X-Correlator values is stateful behaviour, and many API providers won't do this if the request is anyway idempotent.

I can see that there may be cases where the request is not idempotent and the only way to tell between two separate transactions and a duplicated transaction might be such an X-Correlator header, In such a case, the X-Correlator header should indeed be mandatory and duplicates detected and rejected, but this should decided on a case by case basis, and not be mandatory for all API paths.

Incidentally - since 2012, RFC 6648 has deprecated the use of the "X-" prefix for custom headers. It might be worth taking this into account for design guidelines being developed in 2022.